### PR TITLE
Configurable dungeon seed

### LINKED ISSUE #3909: fix seedcracking via usage of dungeons
**Is your feature request related to a problem? Please describe.**
Players using https://github.com/KaptainWutax/SeedCracker can get the seed of any world with a few or even a single dungeon (the details of this are described in this video: https://www.youtube.com/watch?v=8CKh4x4iK38)

**Describe the solution you'd like**
A custom seed that one can apply, basically like the structure seeds in spigot.yml but for populators such as dungeons

**Describe alternatives you've considered**
Removing dungeons? Not really any valid alternatives

### PR DESCRIPTION
**Pull request by BillyGalbreath**
Fixes #3909

### CODE REVIEW

**Comment by KaptainWutax** [comment]
A quick disclaimer that this will cause every single dungeon in the world will be exactly the same (same size, same floor, same loot, same mob type) and around 10% of world seeds will contain no dungeons whatsoever as a side effect (with all 8 attempts being too high in the sky, it's a simple binomial dist with n=8, p=0.25).

**Comment by Proximyst** [comment]
I haven't been following the discussion entirely myself, but I'm fairly sure I've yet to see you provide any kind of input to fix the issues you're describing. Do you have anything to offer there?

**Comment by KaptainWutax** [comment]
I suggest adding a salt like with structures, it should counter a lot of the simpler attacks-- though it will still be possible to use a multi-chunk approach to eliminate salt dependency and continue to abuse dungeons. 

For seed cracking in general, I don't have a proper solution laid out that I am 100% confident will work but I'm working on it. I'm just giving y'all a heads-up since calling this is a fix is a stretch... I find it a little frustrating that one can be this oblivious to their own patch. This will cause major issues for normal casual gameplay if it gets merged. 

The best I can do now is help with the analysis instead of waiting for someone to tweak the seed cracking tools to match those restrictions and have this entire mess start over again. If I can't be trusted to be helpful it that regard though I will gladly step down and figure it out on my own.

**Comment by Proximyst** [comment]
I'm sure your frustrations can be solved by providing a simpler to follow explanation (graph the math, show us where the logic went wrong, etc.) for us, both Billy and anyone reading & participating in the discussion. Apparently, Billy also did not find the 10% having no dungeons to be true in his testing, but I cannot really say more as I haven't done any investigating myself.

**Comment by KaptainWutax** [comment]
Pretty obvious-- same seed = same generation. If you set the decorator to the same seed for every chunk, you will get the same coordinates to get spit out. If you set the feature to the same seed for every chunk, you get the same floor, loot, mob type to get spit out. That 10% comes from the fact that decorator does nextInt(256) to determine dungeon height of 8 trials. I just calculated the probability of 8 trials spawning a dungeon above y64-- which will be consistent across the entire world.

File: Spigot-Server-Patches/0546-Configurable-dungeon-seed.patch [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
++    private Random random; // Paper
+ 
+     public WorldGenDecoratorDungeon(Codec<WorldGenDecoratorDungeonConfiguration> codec) {
+         super(codec);
+@@ -14,10 +15,16 @@ public class WorldGenDecoratorDungeon extends WorldGenDecorator<WorldGenDecorato
+     public Stream<BlockPosition> a(GeneratorAccess generatoraccess, ChunkGenerator chunkgenerator, Random random, WorldGenDecoratorDungeonConfiguration worldgendecoratordungeonconfiguration, BlockPosition blockposition) {
+         int i = worldgendecoratordungeonconfiguration.b;
+ 
++        if (this.random == null) {
++            int seed = com.destroystokyo.paper.PaperConfig.dungeonSeed;
```
Comment (aikar) [inline Spigot-Server-Patches/0546-Configurable-dungeon-seed.patch:38]: This should be stored on world for the instance and then just create local as world version or existing local without any field additions as this instance is concurrently accessed and different worlds can have different seeds

**Review by aikar — CHANGES_REQUESTED** [review]
By not recreating instance this will solve the concern so that matter is closed

Just need to improve impl to store on world

**Comment by KaptainWutax** [comment]
Making dungeons non-deterministic is a fix I suppose. Small suggestion would be to replace the dungeon-seed entry in the config to some boolean "random-dungeons" since the seed doesn't have any effect anymore-- it's pretty much just a new Random(). Maybe you could make use of the world rng? Anyway, looks good. 👍

**Comment by aikar** [comment]
Hmm that's a good point, this should simply behave just like every other seed and be base seed + salt seed for dungeons, then it takes in xz right?

**Comment by KaptainWutax** [comment]
Yea, I think a salt is the best solution if you wanna keep it deterministic.

**Comment by BillyGalbreath** [comment]
Closing this in favor of leaf's patch

**Pull request closed without merging**